pkg/fswallet/fswallet_ethaddr.go

```diff
@@ -0,0 +1,174 @@
+// Copyright © 2023 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fswallet
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/hyperledger/firefly-common/pkg/i18n"
+	"github.com/hyperledger/firefly-signer/internal/signermsgs"
+	"github.com/hyperledger/firefly-signer/pkg/eip712"
+	"github.com/hyperledger/firefly-signer/pkg/ethsigner"
+	"github.com/hyperledger/firefly-signer/pkg/ethtypes"
+	"github.com/hyperledger/firefly-signer/pkg/keystorev3"
+	"github.com/hyperledger/firefly-signer/pkg/secp256k1"
+)
+
+type SyncAddressCallback func(context.Context, ethtypes.Address0xHex) error
+
+// This is a wrapper on the WalletStrIDs capability, that requires all lookups to keys
+// to be ethereum addresses, and adds some other ethereum specific functionality.
+//
+// Originally this library only provided this utility, and later we made the concept
+// of an address more generic.
+type Wallet interface {
+	ethsigner.WalletTypedData
+	GetWalletFile(ctx context.Context, addr ethtypes.Address0xHex) (keystorev3.WalletFile, error)
+	SetSyncAddressCallback(SyncAddressCallback)
+	AddListener(listener chan<- ethtypes.Address0xHex)
+}
+
+type walletEthAddr struct {
+	gw WalletGeneric
+}
+
+func NewFilesystemWallet(ctx context.Context, conf *Config, initialListeners ...chan<- ethtypes.Address0xHex) (ww Wallet, err error) {
+	gw, err := NewFilesystemWalletGeneric(ctx, &ConfigGeneric{
+		Config: *conf,
+		WalletFileValidator: func(ctx context.Context, addrString string, kv3 keystorev3.WalletFile) error {
+			addr, err := ethtypes.NewAddress(addrString)
+			if err == nil {
+				keypair := kv3.KeyPair()
+				if keypair.Address != *addr {
+					err = i18n.NewError(ctx, signermsgs.MsgAddressMismatch, keypair.Address, addr)
```

Comment:
The reason I mentioned the comment above is because you are using that `KeyPair()` API to validate the address here - I think as part of this PR there should be a KeyPairGeneric construct interface which has three methods `GetPubKey` , `GetPrivKey` and `GetAddr`

Reply:
Looking at the code a bit closer, this will result in a lot of changes but is needed to plugin a new type of signing curve in the future

Reply:
Agreed that plugging into the other package `keystorev3` a new form of cryptography would be a lot of work. I haven't proposed that here. Instead (per my comment above) users of this just access the private key from the wallet file using `.PrivateKey()` and can consider the `KeyPair()` function as "historical cruft".